File: @kres/kres.m
function [ K ] = kres( NS, P )
%KRES Constructor of KRES object, given netspec and parameters.
%   A kres object has the following fields:
%       nodeTree - node tree of node labels (from NS)
%       nodes - flat cell array of node labels, in order
%       w - weight matrix
%       wCHL - binary matrix of learnable weights
%       wExemp - binary matrix of exemplar weights
%       bias - array of node biases

if nargin < 2,
    error('KRES:kres:kres', 'Not enough arguments to kres constructor!');
end

% copy in nodeTree
K.nodeTree = NS.nodes;

% flatten nodeTree->nodes
K.nodes = tree2leaves(K.nodeTree);

% this will be a list of connections, mostly for the benefit of logging
K.cons = [];

% construct empty weight matrix (and wCHL/wExemp matricies)
% matricies are the size of K.nodes squared
K.w = zeros(length(K.nodes));
K.wCHL = zeros(length(K.nodes));
K.wExemp = zeros(length(K.nodes));

% activation matricies
K.minusAct = [];
K.plusAct = [];

% stuff for exemplar nodes
% first, a list of indexes into the node list of future exemplar nodes
K.exempNodeIdxs = [];
K.exempActivated = 0; % count
% delta matrixes 
K.exempCHL = zeros(size(K.wCHL)); % CHL connections for a reserved exemplar node
K.exempExemp = zeros(size(K.wExemp)); % Exemplar-update connections for a reserved exemplar node

% do fancy algorithm to fill in weight matrix

% iterate over connections
for c = 1:length(NS.cons),
    % foreach connection, expand the to and from
    toList = tree2leaves(subtree(K.nodeTree, NS.cons(c).to));
    fromList = tree2leaves(subtree(K.nodeTree, NS.cons(c).from));
    
    % if one of the lists is 'E', we'll want to keep a copy of that for
    % later.
    if strcmp(NS.cons(c).to, 'E'),
        K.exempNodeIdxs = getNodeIdx(K, toList);
    elseif strcmp(NS.cons(c).from, 'E'),
        K.exempNodeIdxs = getNodeIdx(K, fromList);
    end
    
    if strcmp(NS.cons(c).spread, 'full'),
        % we'll fill in all pairs of the paired lists
        for toNode = 1:length(toList),
            for fromNode = 1:length(fromList),
                % we've got indexes in toList and fromList, but we need
                % indexes into K.nodes
                toNodeIdx = getNodeIdx(K, toList(toNode));
                fromNodeIdx = getNodeIdx(K, fromList(fromNode));
                
                % no self-connections!
                if (toNodeIdx ~= fromNodeIdx),
                    % store the connections for use by logging
                    K.cons(length(K.cons)+1).from = char(fromList(fromNode));
                    K.cons(length(K.cons)).to = char(toList(toNode));

                    setWeights(fromNodeIdx, toNodeIdx);
                end
            end
        end
        
    else % spread eq 121
        % just a bijection
        if (length(toList) ~= length(fromList)),
            error('KRES:kres:kres', 'Connection #%d is one-to-one, but sizes are different!', c);
        end
        
        for nodeIdx = 1:length(toList),
            % again, we've got indexes in toList etc, need into K.nodes
            toNodeIdx = getNodeIdx(K, toList(nodeIdx));
            fromNodeIdx = getNodeIdx(K, fromList(nodeIdx));
            
            % no self-connections!
            if (toNodeIdx ~= fromNodeIdx),
                % store the connections for use by logging
                K.cons(length(K.cons)+1).from = char(fromList(nodeIdx));
                K.cons(length(K.cons)).to = char(toList(nodeIdx));

                setWeights(fromNodeIdx, toNodeIdx);
            end
        end
    end
    
    %disp(c);
    %disp(toList);
    %disp(fromList);
end

% construct bias array
K.bias = zeros(size(K.nodes));
for b = 1:max(size(NS.bias)),
    % foreach bias, expand the node
    nodeList = tree2leaves(subtree(K.nodeTree, NS.bias(b).node));
    
    for nodeIdx = 1:length(nodeList),
        idx = getNodeIdx(K, nodeList(nodeIdx));
        K.bias(idx) = NS.bias(b).value;
    end
end


% stuff for logging
K.log.ACoherence = 0;
% use minusAct and plusAct for activations
K.log.ASSE = 0;
K.log.Coherence = 0;
K.log.DW = 0;
K.log.Input = [];
K.log.Outputs = [];
K.log.SettleTime = 0;
K.log.SSE = 0;
% weights we got!


% all constructed -- make it a class and we're done!
K = class(K, 'kres');






% this is a *nested* function, not a *subfunction*, for scoping purposes
    function setWeights(fromN, toN)
        % set w matrix
        % value is conditional
        switch NS.cons(c).init
            case 'in'
                wval = getInWeight(P);
            case 'ex'
                wval = getExWeight(P);
            case 'rand'
                wval = rand(1)*getRandWeight(P)*2 - getRandWeight(P);
            otherwise
                if isnumeric(NS.cons(c).init),
                    wval = NS.cons(c).init;
                else
                    wval = 0;
                end
        end
        
        if strcmp(NS.cons(c).type, 'exemp'),
            % actually, don't use wval now...
            wval = 0;
            % instead, add this to exempExemp
            K.exempExemp(fromN,toN) = 1;
            K.exempExemp(toN,fromN) = 1;
        end
        
        if strcmp(NS.cons(c).type, 'exempchl'),
            % for CHL connections to exemplar nodes, put wval into the
            % exempCHL matrix, but then reset to 0, so it doesn't go into
            % the weight matrix yet.
            K.exempCHL(fromN,toN) = wval;
            K.exempCHL(toN,fromN) = wval;
            wval = 0;
        end
                
        K.w(fromN,toN) = wval;
        K.w(toN,fromN) = wval;

        % if type eq chl, set wCHL
        if strcmp(NS.cons(c).type, 'chl'),
            K.wCHL(fromN,toN) = 1;
            K.wCHL(toN,fromN) = 1;
        end

        % wExemp is set only when exemplar nodes are used
    end

    

end % main function

File: R/kres.R
# S4 class for KRES model.
# 
# Author: Harlan
###############################################################################

setClass("kres",
    representation(nodeTree="list", nodes="character", cons="matrix",
        w="matrix", wChl="matrix", wExemp="matrix",
        minusAct="matrix", plusAct="matrix", 
        exempNodeIdx="numeric", exempActivated="numeric",
        exempChl="matrix", exempExemp="matrix",
        bias="matrix"))

setMethod("initialize", "kres", function(.Object, nodeTree, params) {
      # the strategy will be to build local copies of the various matrix
      # objects, then copy them all in at the end...
        
      # flatten nodeTree to nodes
      nodes <- unlist(nodeTree$nodes)
      nNodes <- length(nodes)
      
      # weight matrixes
      # row and column names, oh boy!
      w <- matrix(data=0, nrow=nNodes, ncol=nNodes, dimnames=list(nodes, nodes))
      wChl <- w 	# learnable weights
      wExemp <- w	# exemplar weights
      exempChl <- w	# learnable exemplar initial weight values, to be copied in later
      exempExemp <- w # fixed exemplar connections to input
      # cons matrix for logging
      cons <- w		# all possible connections
      
      # nested functions have to be defined before use. 
      
      # For a single connection, build representations
      buildCons <- function(con) {       
        # get the list of connected nodes
        to <- subtree(nodeTree, con["to"])
        from <- subtree(nodeTree, con["from"])
        
        print(sprintf('%s - %s', con["to"], con["from"]));
        
        # store exemplar nodes for later
        
        #browser()
        
        # now build the connections
        # full
        if (con["spread"] == "full") {
          outer(to, from, function(x, y) mapply(buildConsPair, x, y, list(con)));
        }
        else # one-to-one
        {
          mapply(buildConsPair, to, from, MoreArgs=list(con))
        }
      } # end buildCons
      
      # mapply requires MoreArgs to be a list, so con is always a list of 
      # length 1. Best to just convert back the vector here.
      buildConsPair <- function(a, b, con) {
        if (is.list(con))
          con <- con[[1]];

        if (a != b) {
          cons[a, b] <<- 1
          cons[b, a] <<- 1
          
          setWeights(a, b, con)
        }
        
        0 # to prevent outer() and mapply() from choking on NULL
      }
      
      setWeights <- function(a, b, con) {
        # first, figure out the weight value
        # switch() requires either integer indexes or strings, and we want
        #  strings here, so force character, but test later for number
        wval <- switch(as.character(con["init"]),
            inh = params$inWeight,
            exh = params$exWeight,
            rand = runif(1, -params$randWeight, params$randWeight),
            #otherwise:
            ifelse(is.numeric(con["init"]), con["init"], 0))
        
#        print(sprintf('wval = %f', wval))
        
        if (con["type"] == "exemp") {
          # actually, don't use wval for fixed exemplar nodes
          wval <- 0
          # instead, add this to exempExemp
          exempExemp[a,b] <<- 1
          exempExemp[b,a] <<- 1
        }
        
        if (con["type"] == "exempchl") {
          # for CHL connections to exemplar nodes, put wval into the
          # exempChl matrix, but then reset to 0, so it doesn't go into
          # the weight matrix yet.
          exempChl[a,b] <<- wval
          exempChl[b,a] <<- wval
          wval <- 0
        }
        
        # set the weights themselves
        w[a,b] <<- wval
        w[b,a] <<- wval
        
        if (con["type"] == "chl") {
          wChl[a,b] <<- 1
          wChl[b,a] <<- 1
        }
        
       
      }
      
      # back to the main body of kres::initialize()

      # activation matrixes
      
      # exemplar node stuff
      
      # construct weight matrix from the node tree
      lapply(nodeTree$cons, buildCons) # lots of side effects here
      
      # construct bias array
      bias <- matrix(data=0, nrow=nNodes, ncol=1, dimnames=list(nodes))
      
      # logging 

      
      .Object@w <- w
      .Object@wChl <- wChl
      .Object@wExemp <- wExemp
      .Object@exempChl <- exempChl
      .Object@exempExemp <- exempExemp
      .Object@nodeTree <- nodeTree  # train and test will need a copy
      .Object@nodes <- nodes
      .Object@bias <- bias
  
      .Object
    })

setGeneric("converge", function(.K, Xin, params) standardGeneric("converge"))

setMethod("converge", "kres", function(.K, Xin, params) {
      harmony <- 0
      dHarmony <- 1
      
      # net_input starts out as a vector of 0s
      net_input <- Xin * 0
      adj_input <- Xin * 0
      
      cycles <- 0
      
      # bleah, need accessors
      # do this out of the loop for speed, maybe
      gain <- params$gain
      alpha <- params$alpha
      bias <- .K@bias
      w <- .K@w
      
      
      # while harmony is high and we haven't done a ridiculous number of cycles
      while ((dHarmony > 0.00001) && (cycles < 1000000))
      { 
        # calculate total_input
        total_input <- net_input + Xin + t(bias)
      
        # do gain thing
        adj_input <- adj_input + (total_input - adj_input) / gain; 
      
        # calculate activation
        act <- 1 / (1 + exp(-alpha * adj_input))
        
        # calculate net_input
#        browser()
        net_input <- act %*% w
        
        # calculate harmony
        newHarmony <- sum(act * net_input);
        dHarmony <- abs(harmony - newHarmony);
        harmony <- newHarmony;
        
        cycles <- cycles + 1;
      }
          
      # return a list with the post-convergence activation and the convergence time in 
      # steps
      return(list(act=act, cycles=cycles))
    })


setGeneric("test", function(.K, Xin, Xout, params) standardGeneric("test"))

setMethod("test", "kres", function(.K, Xin, Xout, params) {
      # OK! Here's how this will work. First, converge with just Xin as input,
      # and save the activations. 
          
      convRet <- converge(.K, Xin, params)
      
      # Now, figure out the ensemble's prediction and error. 
          
      # Use the node tree in K to get the output nodes. Use those to figure out
      # which elements of K.minusAct belong in the output vector.
      # Then, do LCR on that to get.
      
#          outList = tree2leaves(subtree(K.nodeTree, 'O'));
#      outAct = zeros(size(outList));
#      for i = 1:length(outAct),
#      outAct(i) = K.minusAct(getNodeIdx(K, outList(i)));
#      end
#      
#      K.log.Outputs = outAct ./ sum(outAct);
#      out = K.log.Outputs;
#      
#      # and for logging...
#          K.log.Input = Xin + Xout;
      
      
      
    })


#SUBTREE Get the subtree of T rooted at head.
subtree <- function(T, head)
{
  if (is.null(names(T))) {
    # it's a list of leaves, so return a leaf or null
    if (head %in% T) { ret <- head } else { ret <- NULL }
  }
  else if (head %in% names(T)) {
    # it's at this level, so return it
    ret <- T[head]
  }
  else {
    # it's not here, so recurse
    ret <- unlist(lapply(T, function(x) subtree(x, head)))
           # unlist gets rid of the null calls!
  }
  ret
}